### ISSUE TITLE
Extracting Private Keys From Keystore?

### ISSUE THREAD

**Issue by keithpiong** [Boilertalk/Web3.swift#28]
if only there's a simpler way to extract private keys directly from keystore

**Comment by Ybrin** [comment]
@pixelmatrix Do you know the keystore implementation? I never worked with it directly so I don't know exactly what we would need for it.    
Some questions I have are:

* Is the actual encryption available in [CryptoSwift](https://github.com/krzyzanowskim/CryptoSwift) or do we need another dependency/our own implementation of it?
* Same question for the key derivation function they use
* If we need to implement them on our own, is it worth the effort/can we do it in reasonable amount of time compared to what it delivers?

Ok that's it for now...If you haven't used it before I will try and search some specifications.

**Comment by pixelmatrix** [comment]
Hey @Ybrin, i'm not really sure about the keystore. Our keys are all managed server side so we avoid needing to use that. Is this referring to iOS Keychain or something else?

**Comment by Ybrin** [comment]
@pixelmatrix No I think he is talking about the files geth generates (also MEW I think?) if you create a new wallet.

This one: https://medium.com/@julien.m./what-is-an-ethereum-keystore-file-86c8c5917b97

**Comment by pixelmatrix** [comment]
Interesting… it doesn't seem like there's a json rpc method for that, so you'd need direct access to the node's filesystem somehow. Assuming you had the json file, is the request for being able to parse that file into an `EthereumPrivateKey` object?

**Comment by keithpiong** [comment]
Yes, @Ybrin is right. I have been using Geth to create wallet and I am able to use the export and import key function. And yes, the keystore are JSON encoded. Can be found with the link below:

https://github.com/ethereum/go-ethereum/wiki/Mobile:-Account-management

I understand that if I were to keep the keystore file, it is much more secure instead of keeping my private key. However, it gets a bit complicated when it comes to signing & sending transactions because signing the transaction requires the private key (as hex string) which is encrypted inside the JSON encoded keystore (AES - 128 - ctr). 

So the clearer question to ask is:
During the signing process, is there a simpler method to decrypt the keystore file (with relevant passphrase of course) and extract the private key? Can CryptoSwift able to handle this decryption process alone?

**Comment by Florian-S-A-W** [comment]
I think you are not missing anything in the JS implementation. Note that recover returns the private key as a buffer, I suggest you take a look at: https://nodejs.org/api/buffer.html#buffer_class_buffer

**Comment by Ybrin** [comment]
@keithpiong Check out https://github.com/Boilertalk/Keystore.swift

**Comment by Ybrin** [comment]
Run `pod repo update` and then again `pod install`

**Comment by vickyiosdev** [comment]
**let privateKey =  try?  (key.privateKey(password: password))**

For above mentioned line of code to get private key it is taking almost **6minutes**.
please help me out here.

**Comment by Ybrin** [comment]
@vickyiosdev You'll have to production compile your app. It will greatly
reduced the time it takes to encrypt/decrypt. On a normal iPhone it should
be just a few seconds (1-5). This is an issue with the underlying hashing
and encryption libraries we use. They are really inefficient if compiled
without the optimization flags...

**Comment by vickyiosdev** [comment]
@Ybrin Thanks,
In debug mode should we change any flags in Xcode.
In release mode app is crashing.
please help me out here.

**Comment by Ybrin** [comment]
You could manually add maximum optimization flags to the debug build. But
you should really fix your release build at some point. Is the crash in any
way related to Web3 or Keystore? I need the logs to help you...

**Comment by vickyiosdev** [comment]
> You could manually add maximum optimization flags to the debug build. But you should really fix your release build at some point. Is the crash in any way related to Web3 or Keystore? I need the logs to help you...
> […](#)

Fixed Crash issues.
@Ybrin Could you please provide clear idea on optimization flags, which flags i need to update in Xcode.
Thanks.

**Comment by vickyiosdev** [comment]
@Ybrin Unable to show activity indicator while extracting private key.
UI is getting freeze.
Note: In main thread I'm enabling activity indicator.

Could you please help me out here.
